packages/babel-core/src/transformation/normalize-opts.js

```diff
@@ -3,19 +3,41 @@
 import type { ResolvedConfig } from "../config";
 
 export default function normalizeOptions(config: ResolvedConfig): {} {
-  const options = Object.assign({}, config.options, {
+  const opts = config.options;
+
+  const options = Object.assign({}, opts, {
     parserOpts: Object.assign(
       {
-        sourceType: config.options.sourceType,
-        sourceFileName: config.options.filename,
+        sourceType: opts.sourceType,
+        sourceFileName: opts.filename || "unknown",
         plugins: [],
       },
-      config.options.parserOpts,
+      opts.parserOpts,
+    ),
+    generatorOpts: Object.assign(
```

Comment:
I think we can use object spread in these places if you free it's more readable

Reply:
I think I'll leave this since it was already using `Object.assign`, but fair point. Will keep that in mind for next time.